### DISCUSSION TITLE
Unable to set up Topic Metrics

### DISCUSSION THREAD

**Discussion by oscgonfer** [emqx/emqx#13272] [Q&A] [answered]
Hi all,

We have a running instance of EMQX v5.3.2, deployed on an ubuntu 22.04 LTS Machine via Docker.

We are trying to set up topic metrics, but we are not getting it right, or at least seems we are missing a step or detail.

Below, you can see our various attempts to refine the topic via different methods, but we have not managed to get any stats on them.

![imagen](https://github.com/emqx/emqx/assets/13301321/5f804188-5bfe-48d0-a438-c4925331a9f7)

What are we missing?

Thank you in advance!

**Comment by keynslug** [comment, marked as answer]
Hi @oscgonfer,

I haven't managed to reproduce. To clarify this is what I did.
1. Run EMQX in a docker container, with debug logging for better insights.
    ```
    $ docker run -p 18083:18083 -p 1883:1883 -e EMQX_LOG__CONSOLE_HANDLER__LEVEL=debug --name emqx-dev docker.io/emqx/emqx:5.3.2
    ```
2. Enable metrics for topic `topic/42`.
3. Run MQTT client in a separate shell few times.
    ```
    $ mosquitto_pub -h '127.0.0.1' -p 1883 -t 'topic/42' -m 'PUB'
    ```

This is what I'm getting.
![image](https://github.com/emqx/emqx/assets/1148366/e34a0ff1-6af3-4ec5-9bfa-e1dcb84d806e)

I'd probably double-check that the clients are in fact talking with EMQX as expected.

If debug logging is a no-go for you, you could as well set up a simple rule with an action that prints evaluation results to the stdout, i.e. with the following RuleSQL.
```
SELECT topic
FROM "#"
```
This might help to understand what's happening.

**Reply by oscgonfer** [reply]
I think I see what's happening. I just published to any of the "bare" topics there, but there are no metrics on the subtopics themselves.

This makes the counter increase on `device`:

```
mosquitto_pub -h 127.0.0.1 -t 'device' -m "hello"
```

But this doesn't:

```
mosquitto_pub -h 127.0.0.1 -t 'device/device' -m "hello"
```

Which makes it a bit unuseful for the topic metrics. I understand that wildcards are more complex, but at least subtopics would be quite a useful feature.

**Reply by oscgonfer** [reply]
@timcowlishaw check for our particular case

**Reply by keynslug** [reply]
I see, thanks for confirming.

> I understand that wildcards are more complex, but at least subtopics would be quite a useful feature.

I agree, but from the point-of-view of the MQTT spec there's no such concept as subtopics. So this likely falls under the "aggregated topic metrics by filters" umbrella.

**Comment by oscgonfer** [comment]
Hi,

Thanks! Sorry for my lack of precission.

Just to clarify I am not missing something. Can you clarify what you mean
by?

> "aggregated topic metrics by filters" umbrella

I just wanted to make sure we are not missing features.

Best,




Óscar
